File: backend/src/main/java/com/programmers/pcquotation/domain/estimate/entity/Estimate.java
package com.programmers.pcquotation.domain.estimate.entity;

import com.programmers.pcquotation.domain.comment.emtity.Comment;
import com.programmers.pcquotation.domain.estimaterequest.entity.EstimateRequest;
import com.programmers.pcquotation.domain.seller.entitiy.Seller;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor
public class Estimate {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	public Integer id;

	@NotNull
	@ManyToOne
	public EstimateRequest estimateRequest;

	@NotNull
	@ManyToOne
	public Seller seller;

	@Setter
	public  Integer totalPrice;

	@Setter
	@OneToMany(mappedBy = "estimate", cascade = CascadeType.ALL, orphanRemoval = true)
	public List<EstimateComponent> estimateComponents = new ArrayList<>();

	public  LocalDateTime createDate;

	@OneToMany(mappedBy = "estimate", cascade = CascadeType.ALL, orphanRemoval = true)
	public List<Comment> comments;

	@Builder
	public Estimate(EstimateRequest estimateRequest, Seller seller, Integer totalPrice,
		List<EstimateComponent> estimateComponents) {
		this.estimateRequest = estimateRequest;
		this.seller = seller;
		this.totalPrice = totalPrice;
		this.createDate = LocalDateTime.now();
		this.estimateComponents = estimateComponents;
	}

	public void addEstimateComponent(EstimateComponent component) {
		this.estimateComponents.add(component);
		component.setEstimate(this);
	}

}

File: backend/src/main/java/com/programmers/pcquotation/domain/estimate/entity/Estimate.kt
package com.programmers.pcquotation.domain.estimate.entity

import com.programmers.pcquotation.domain.comment.emtity.Comment
import com.programmers.pcquotation.domain.estimaterequest.entity.EstimateRequest
import com.programmers.pcquotation.domain.seller.entitiy.Seller
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
class Estimate(
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	var id: Int = 0,

	@ManyToOne
	var estimateRequest: EstimateRequest,

	@ManyToOne
	var seller: Seller,

	var totalPrice: Int = 0,

	@OneToMany(mappedBy = "estimate", cascade = [CascadeType.ALL], orphanRemoval = true)
	private val _estimateComponents: MutableList<EstimateComponent> = mutableListOf(),

	var createDate: LocalDateTime = LocalDateTime.now(),

	@OneToMany(mappedBy = "estimate", cascade = [CascadeType.ALL], orphanRemoval = true)
	var comments: List<Comment> = ArrayList(),

	@Column(nullable = false)
	var isAdopted: Boolean = false
) {
	val estimateComponents: List<EstimateComponent>
		get() = _estimateComponents.toList()

	fun addEstimateComponents(components: List<EstimateComponent>) {
		components.forEach { component ->
			addEstimateComponent(component)
		}
	}

	fun deleteEstimateComponents() {
		_estimateComponents.clear()
	}

	private fun addEstimateComponent(component: EstimateComponent) {
		_estimateComponents.add(component)
		component.estimate = this
	}
}